components/formatters/src/duration/duration-formatter.ts

```diff
@@ -0,0 +1,249 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  DtFormattedValue,
+  FormattedData,
+  NO_DATA,
+  SourceData,
+} from '../formatted-value';
+import { DtTimeUnit } from '../unit';
+
+enum DurationMode {
+  DEFAULT = 'DEFAULT',
+  PRECISE = 'PRECISE',
+}
+
+// tslint:disable: no-magic-numbers
+// Factorials needed for converting milliseconds to other time units
+const CONVERSION_FACTORS_TO_MS = new Map<DtTimeUnit, number>([
+  [DtTimeUnit.YEAR, 12 * 30.41666 * 24 * 60 * 60 * 1000],
+  [DtTimeUnit.MONTH, 30.41666 * 24 * 60 * 60 * 1000],
+  [DtTimeUnit.DAY, 24 * 60 * 60 * 1000],
+  [DtTimeUnit.HOUR, 60 * 60 * 1000],
+  [DtTimeUnit.MINUTE, 60 * 1000],
+  [DtTimeUnit.SECOND, 1000],
+  [DtTimeUnit.MILLISECOND, 1],
+  [DtTimeUnit.MICROSECOND, 0.001],
+  [DtTimeUnit.NANOSECOND, 0.000001],
+]);
+
+const CONVERSIONUNITS = 3;
+
+/**
+ * Formats a numeric value to a duration
+ * @param duration numeric time value
+ * @param formatMethod the formatting method
+ * @param outputUnit dtTimeUnit | undefined value describing the unit to which it should format e.g to seconds
+ * @param inputUnit dtTimeUnit value describing which unit the duration is in
+ */
+export function formatDuration(
+  duration: number,
+  formatMethod: string | undefined,
+  outputUnit: DtTimeUnit | undefined,
+  inputUnit: DtTimeUnit = DtTimeUnit.MILLISECOND,
+): DtFormattedValue | string {
+  const inputData: SourceData = {
+    input: duration,
+    unit: inputUnit,
+  };
+  let formattedData: FormattedData;
+  let result: Map<DtTimeUnit, string> | undefined;
+
+  if (duration <= 0 && formatMethod !== DurationMode.PRECISE) {
+    // Smaller than 0 duration impossible
+    return new DtFormattedValue(inputData, {
+      transformedValue: inputData.input,
+      displayValue: '< 1',
+      displayUnit: inputUnit,
+      displayWhiteSpace: false,
+    });
+  } else {
+    if (!formatMethod) {
+      formatMethod = DurationMode.DEFAULT;
+    }
+    if (formatMethod === DurationMode.PRECISE) {
+      result = transformResultWithOutputUnit(
+        duration,
+        inputUnit,
+        outputUnit,
+        formatMethod,
+      );
+    } else if (formatMethod === DurationMode.DEFAULT) {
+      if (outputUnit) {
+        result = transformResultWithOutputUnit(
+          duration,
+          inputUnit,
+          outputUnit,
+          formatMethod,
+        );
+      } else {
+        result = transformResult(duration, inputUnit, formatMethod);
+      }
+    } else {
+      result = transformResult(duration, inputUnit, formatMethod);
+    }
```

Comment:
There are a lot of if else switches also with redundant branches - we can definitely simplify this - let's sync on that one in a more direct channel 😄